c10/core/TensorImpl.h

```diff
@@ -850,14 +864,24 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   // `allow_tensor_metadata_change_` to false would prevent those changes from happening and is
   // undesirable.
   virtual c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach() const {
-    auto impl = c10::make_intrusive<TensorImpl>(Storage(storage()), type_id(), is_variable());
-    impl->set_sizes_and_strides(sizes(), strides());
-    impl->storage_offset_ = storage_offset_;
-    impl->is_wrapped_number_ = is_wrapped_number_;
-    impl->reserved_ = reserved_;
-    impl->refresh_numel();
-    impl->refresh_contiguous();
-    return impl;
+    if (!opaque_handle_) {
+      auto impl = c10::make_intrusive<TensorImpl>(Storage(storage()), type_id(), is_variable());
+      impl->set_sizes_and_strides(sizes(), strides());
+      impl->storage_offset_ = storage_offset_;
+      impl->is_wrapped_number_ = is_wrapped_number_;
+      impl->reserved_ = reserved_;
+      impl->refresh_numel();
+      impl->refresh_contiguous();
+      return impl;
+    } else {
+      auto impl = c10::make_intrusive<TensorImpl>(type_id(), dtype(), is_variable(), opaque_handle_, sizes());
```

Comment:
I think this is problematic. `ideep::tensor` keeps another copy of its sizes, but you are sharing the underlying `ideep` tensor when you do this. I think we also need to do a shallow copy of ideep tensor as well. Otherwise, you can get into trouble this way:

```
# let x be an MKLDNN tensor
y = x.detach()
x.resize_(400)
# what's the size of y?
```

The resize operation will have changed the size of `ideep::tensor`, but the metadata on `y` (which was shallow copied, remember) is now out of date. So you'll violate the invariant that tensor size and ideep tensor size match.

Reply:
An alternate resolution to this problem, is that if the opaque handle is placed in storage, we never need to shallow copy it and so we can just treat it effectively as a unique pointer, solving @dzhulgakov's earlier comment.

Reply:
> The resize operation will have changed the size of `ideep::tensor`, but the metadata on `y` (which was shallow copied, remember) is now out of date. So you'll violate the invariant that tensor size and ideep tensor size match.

@ezyang The tensor size and ideep tensor size should be guaranteed to be consistent in their respective kernels, e.g. resize_mkldnn_ which I haven't added yet. Now when you call `x.resize_`, you will get an exception.

Reply:
> An alternate resolution to this problem, is that if the opaque handle is placed in storage, we never need to shallow copy it and so we can just treat it effectively as a unique pointer, solving @dzhulgakov's earlier comment.

@ezyang Are you suggesting to add the opaque handle into the `Storage` instead of `TensorImpl`? AFAIK, `Storage` abstracts a typed resize-able ref-countable 1D array on a given device. But the opaque handle has both the memory buffer and the metadata. Putting opaque handle into `Storage` changes its abstraction and complicates its semantics. It doesn't sound a good idea.

Reply:
> The tensor size and ideep tensor size should be guaranteed to be consistent in their respective kernels, e.g. resize_mkldnn_ which I haven't added yet. Now when you call x.resize_, you will get an exception

What I am saying is that, under the current design where a shallow copy of the tensor shares the `ideep::tensor`, it is impossible to guarantee consistency, no matter how hard you try in `resize_mkldnn_`.

Let's consider a hypothetical implementation of `resize_mkldnn_` (pseudocode:

```
void resize_mkldnn_(Tensor input, IntArrayRef new_sizes) {
  ideep::tensor ideep_input itensor_from_mkldnn(input);
  ideep_input.resize(new_sizes, ...);
  input.unsafeGetTensorImpl()->set_sizes(new_sizes, ...);
}
```

Consider what happens if you use this function with the following sample program:

```
void example(Tensor input) {
  new_input = input.shallow_copy_and_detach();
  // new_input shares ideep::tensor with input, but is a separate TensorImpl struct
  resize_mkldnn_(input, new_sizes);
  // now input and ideep::tensor have updated sizes, but new_input has OLD sizes
}
```

There's no way to "fix" `resize_mkldnn_` because `new_input` is not reachable from `input`.

> Are you suggesting to add the opaque handle into the Storage instead of TensorImpl? AFAIK, Storage abstracts a typed resize-able ref-countable 1D array on a given device. But the opaque handle has both the memory buffer and the metadata. Putting opaque handle into Storage changes its abstraction and complicates its semantics. It doesn't sound a good idea.

Right: what you are saying is, storage abstracts the *buffer*, which I totally agree. This is one of the reasons why, in my other comment, I suggested that we shouldn't store an `ideep::tensor` directly; instead we ought to store the pieces that constitute an `ideep::tensor` in their correct spots (buffer inside storage; metadata in tensor) and reconstruct `ideep::tensor` on the fly.

I suppose the other way we can "solve" this problem is to disallow metadata-changing operations on ideep tensors entirely. I don't know if that is a step too far for you though.

Reply:
Well, the essential problem which prevent us from embedded metadata to Pytorch Tensor's metadata is that we have a superset of them. And resize a MKL-DNN tensor shouldn't be allowed since we can't resize a tensor with intricate layout. What I can think of is to transform MKL-DNN tensor back to CPU tensor and do the resizing. @ezyang

Reply:
@4pao Hmm, I think there are two problems here:
1. On the metadata change issue, the "copy-on-write" contract for opaque tensor should be able to solve it. See my previous comments below:
> As an alternative solution, how about we always do "copy-on-write" anytime the metadata are being changed? For example, suppose we have to implement resize_mkldnn_ or any op that might change the metadata inside the opaque handle, we do explicit clone for the metadata.
```
Tensor& resize_mkldnn_(Tensor& input, IntArrayRef new_sizes) {
  ideep::tensor ideep_input = itensor_from_mkldnn(input);
  ideep::tensor ideep_output = ideep_input.shallow_copy(); // not implement yet but can add later
  ideep_output.resize(new_sizes, ...);
  input = new_mkldnn_with_itensor(ideep_output);
  return input;
}
```
2. On whether MKL-DNN tensor should support resizing, I don't know if we really need it now. But semantically we can do an internal conversion to plain format first before resizing, not necessarily converting it back to CPU tensor.